pkg/storage/storagedb.go

```diff
@@ -0,0 +1,293 @@
+		cache.WithExpireAfterWrite(options.EvictionTTL),
+	}
+	if options.UseDisk() {
+		cacheOptions = append(cacheOptions, cache.WithRemovalListener(storageDB.OnCacheRemovalCallback))
+	}
+	cacheDb := cache.New(cacheOptions...)
+	storageDB.cache = cacheDb
+
+	if options.UseDisk() {
+		// if the path exists we create a random temporary subfolder
+		if !fileutil.FolderExists(options.DbPath) {
+			return nil, errors.New("folder doesn't exist")
+		}
+		dbpath := filepath.Join(options.DbPath, xid.New().String())
+
+		if err := os.MkdirAll(dbpath, 0644); err != nil {
+			return nil, err
+		}
+		levDb, err := leveldb.OpenFile(dbpath, &opt.Options{})
+		if err != nil {
+			return nil, err
+		}
+		storageDB.dbpath = dbpath
+		storageDB.db = levDb
+	}
+
+	return storageDB, nil
+}
+
+func (s *StorageDB) OnCacheRemovalCallback(key cache.Key, value cache.Value) {
+	if key, ok := value.([]byte); ok {
+		_ = s.db.Delete(key, &opt.WriteOptions{})
+	}
+}
+
+func (s *StorageDB) GetCacheMetrics() (*CacheMetrics, error) {
+	info := &cache.Stats{}
+	s.cache.Stats(info)
+
+	cacheMetrics := &CacheMetrics{
+		HitCount:         info.HitCount,
+		MissCount:        info.MissCount,
+		LoadSuccessCount: info.LoadSuccessCount,
+		LoadErrorCount:   info.LoadErrorCount,
+		TotalLoadTime:    info.TotalLoadTime,
+		EvictionCount:    info.EvictionCount,
+	}
+
+	return cacheMetrics, nil
+}
+
+// SetIDPublicKey sets the correlation ID and publicKey into the cache for further operations.
+func (s *StorageDB) SetIDPublicKey(correlationID, secretKey, publicKey string) error {
+	// If we already have this correlation ID, return.
+	_, found := s.cache.GetIfPresent(correlationID)
+	if found {
+		return errors.New("correlation-id provided already exists")
+	}
+	publicKeyData, err := ParseB64RSAPublicKeyFromPEM(publicKey)
+	if err != nil {
+		return errors.Wrap(err, "could not read public Key")
+	}
+	aesKey := uuid.New().String()[:32]
+
+	ciphertext, err := rsa.EncryptOAEP(sha256.New(), rand.Reader, publicKeyData, []byte(aesKey), []byte(""))
+	if err != nil {
+		return errors.New("could not encrypt event data")
+	}
+
+	data := &CorrelationData{
+		SecretKey:       secretKey,
+		AESKey:          []byte(aesKey),
+		AESKeyEncrypted: base64.StdEncoding.EncodeToString(ciphertext),
+	}
+	s.cache.Put(correlationID, data)
+	return nil
+}
+
+func (s *StorageDB) SetID(ID string) error {
+	data := &CorrelationData{}
+	s.cache.Put(ID, data)
+	return nil
+}
+
+// AddInteraction adds an interaction data to the correlation ID after encrypting
+// it with Public Key for the provided correlation ID.
+func (s *StorageDB) AddInteraction(correlationID string, data []byte) error {
+	item, found := s.cache.GetIfPresent(correlationID)
+	if !found {
+		return errors.New("could not get correlation-id from cache")
+	}
+	value, ok := item.(*CorrelationData)
+	if !ok {
+		return errors.New("invalid correlation-id cache value found")
+	}
+
+	value.Lock()
+	if s.Options.UseDisk() {
```

Comment:
we should not hold lock over the entire duration of this processing and only while using value items